### PR TITLE
Unify context menus

### PR DESCRIPTION
**Pull request by tamland**
This is a start at fixing current context menu code and move it to a global item based system (which will also support arbitrary nesting of menus). We already have this for addons, so this PR simply hooks up this system to allow non-addon menus to be added, and adds Interoperability with the old system.

Also adds context menu support to directory providers and ports a few menus as examples.

Next: port _all_ current context menus to `IContextMenuItem` implementations and drop the old ones. Volunteers?:)

### CODE REVIEW

**Comment by Tolriq** [comment]
@tamland do you think that after this PR it will be possible to extend that so it can be accessed from JSON RPC ?

**Comment by Razzeee** [comment]
Awesome, very good step into the right direction. I really hated all those add(menuItem) scattered everywhere.

**Comment by BigNoid** [comment]
Awesome indeed, this adds a wide range of possibilities.

**Comment by Razzeee** [comment]
@tamland 
please pick https://github.com/Razzeee/xbmc/commit/6112c081ecb2c4aa292f0fb60124912807eb56c2

and consider https://github.com/Razzeee/xbmc/commit/222ae612c4c97dc917a68411c7adc218e7accef5

**Comment by Razzeee** [comment]
Pick https://github.com/xbmc/xbmc/commit/1d4a76f0f4e2d0d4cfd607c67cc80a38787e5464 for the remaining musics info stuff

image
The bottom left menu part is missing for some reason, but might be skin related?  @phil65

**Comment by phil65** [comment]
yes, skin related.
@tamland  latest estuary master already contains homescreen preparations so that the infoscreens mostly look fine.
Thx for this! Really awesome work.

**Comment by Montellese** [comment]
Nice work (haven't looked at all the changes).

Personally I'd prefer the `CTvShowInfo` implementation example to be in its own class and part of the `xbmc/video` directory because that's where it belongs.

I'm also not a huge fan of having a huge list of "interface implementations" hardcoded in the manager initialization. Every implementation should be injected into the manager from some place where it belongs. But that's not easy to do (which is why I haven't solved it yet either in my media importing stuff).

**Comment by da-anda** [comment]
one thing to keep an eye on is that if users modify an item in the info views (like slecting a new poster), that the underlying window needs to be notified to update the info of this item as well or it will show the old artwork/name f.e.

**Comment by tamland** [comment]
@Montellese If it wasn't clear, the implementations here was just proof of concept. I don't intend to have them all in one huge header file! The final connection of implementation the manager however is by design done in a central place atm. This is to keep it simple and straight forward, and it's required to have complete control of ordering here.

@Razzeee I've take the liberty of refactoring the info menus because that turned into way too much boilerplate too fast:) As @Montellese said we should try to find appropriate places to keep the implementations. Keep them coming!

**Comment by Razzeee** [comment]
@tamland 
did you miss https://github.com/xbmc/xbmc/commit/1d4a76f0f4e2d0d4cfd607c67cc80a38787e5464 ? or should I refactor it accordingt to the new way?

**Comment by Montellese** [comment]
@tamland: Ah I didn't realize that order matters. Maybe this should be documented in a comment somewhere near the definition or initialization of `m_items`.

**Comment by Razzeee** [comment]
@tamland 
Pick https://github.com/Razzeee/xbmc/commit/ad5df93e6c20bb4fdaf774d347ed9bfe746138b0

We will also need a xcode and vs sync when we're done with everything.

**Comment by tamland** [comment]
I've added a few more menus. I think I'll stop now, otherwise I'll never get done:)

Still looking for xcode. Anyone? @anaconda?

**Comment by Razzeee** [comment]
Should I look at vs sync again if you say your done for now?

**Comment by tamland** [comment]
Shouldn't be necessary. I've done it manually

**Comment by anaconda** [comment]
@tamland Last 3 commits here add the new files to Xcode:
https://github.com/anaconda/xbmc/commits/9257-xcode

**Comment by tamland** [comment]
Thanks @anaconda!
jenkins build this please

**Comment by tamland** [comment]
jenkins build this please

**Comment by NedScott** [comment]
Is this all "under the hood" or does it need any changes to the wiki? (granted, the wiki isn't entirely up-to-date on context menu options as-is)

**Comment by tamland** [comment]
Mostly 'under the hood', yes, but there's a few smaller adjustments to the some of the current menus. There will likely be more changes before Krypton though, so I would wait with updating the wiki until it 'settle'.

**Comment by NedScott** [comment]
That sounds good. Those will probably be picked up as we start to take new screenshots for everything, due to the new skin :)

File: xbmc/video/windows/GUIWindowVideoNav.cpp [hunk 1]
```diff
@@ -92,10 +92,12 @@ bool CGUIWindowVideoNav::OnAction(const CAction &action)
     CFileItemPtr pItem = m_vecItems->Get(m_viewControl.GetSelectedItem());
     if (pItem->IsParentFolder())
       return false;
-    if (pItem && pItem->GetVideoInfoTag()->m_playCount == 0)
-      return OnContextButton(m_viewControl.GetSelectedItem(),CONTEXT_BUTTON_MARK_WATCHED);
-    if (pItem && pItem->GetVideoInfoTag()->m_playCount > 0)
-      return OnContextButton(m_viewControl.GetSelectedItem(),CONTEXT_BUTTON_MARK_UNWATCHED);
+
+    if (pItem && pItem->HasAddonInfo())
```
Comment (anaconda) [inline xbmc/video/windows/GUIWindowVideoNav.cpp]: `HasVideoInfoTag`? `IsVideoDb`?
See http://forum.kodi.tv/showthread.php?tid=266143

**Pull request merged**